---
title: "Analyzing MeRIP-seq data with PoissonGamma"
author: "Zijie Zhang, Allen Zhu, Chuan He, Mengjie Chen"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('RADAR')`"
abstract: >
  A basic task in the analysis of MeRIP-seq is the detection of differentially 
  methylated loci on the RNA. To analyze the count data of MeRIP-seq, we divide 
  the transcript (concatenated exons of a gene) into consecutive bins and 
  quantify the reads align to each bin. The summarized count data is in table
  format where each row represent a loci and each column is an sample. Statistical
  inference of changes across group, as compared to within-group variability was 
  made for each bin with sufficient count. This vignette explains the use of the package RADAR with
  example data.
  RADAR package version: `r packageVersion("RADAR")`
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_float: true
    fig_width: 8
vignette: >
  %\VignetteIndexEntry{Analyzing MeRIP-seq data with RADAR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=TRUE,
                      dev="png",
                      message=FALSE, error=FALSE, warning=TRUE)
```	


**Note:** if you use RADAR in published research, please cite:

> 
> 
> *Guess What Biology*, **15**:550.
> [](http://dx.doi.org/10.1186/s13059-014-0550-8)

Other Bioconductor packages with similar aims are
[exomePeak](http://bioconductor.org/packages/exomePeak),
[MeTDiff](https://github.com/compgenomics/MeTDiff)
[QNB](https://cran.r-project.org/web/packages/QNB/index.html).

## 1.Quick start

Here we show the most basic steps for a differential methylation
analysis. There are one major steps upstream of radar that generate 
mapped reads in BAM files for each sample, which can be done by varies
choices of softwares such as [Hisat2](https://ccb.jhu.edu/software/hisat2/index.shtml), 
[tophat2](http://www.ccb.jhu.edu/software/tophat/index.shtml) or [Kallisto](https://pachterlab.github.io/kallisto/). 
we will discuss in the sections below assuming BAM file has been obtained. 

This code chunk assumes that you have bam files for sample1-6 `s1.input.bam... s6..input.bam`
and `s1.m6A.bam... s6..m6A.bam` and a vector of group asignment `X = c("Ctl","Ctl","Ctl","Treated","Treated","Treated")`. 

```{r quickStart, eval=FALSE}
radar <- countReads(
  samplenames = c("s1","s2","s3","s4","s5","s6"), 
  gtf = "/path/to/file.gtf",
  bamFolder = "/path/to/bam files",
  modification = "m6A", 
  strandToKeep = "opposite"
)

radar <- normalizeLibrary( radar, X )
radar <- adjustExprLevel( radar )
radar <- filterBins( radar ,minCountsCutOff = 15)
radar <- diffIP(radar)
res <- reportPoissonGammaMerge(radar,cutoff = 0.1)
```

## 2.How to get help for 
Issues and questions about the package should be posted to the [issue report page of the github](https://github.com/scottzijiezhang/RADAR/issues). The issue discussed here can served as public knowledge base.  

## 3.Standard workflow

### a). Mapped BAM file. 
To analyze for MeRIP-seq data, **RADAR** expect a pair of BAM file for each sample. For each pair of BAM file, the naming convention for **RADAR** is `sample.input.bam` for **INPUT** and `sample.m6A.bam` for m6A **IP** or `sample.m1A.bam` for m1A **IP** sample. They key is **INPUT** and **IP** sample have the same prefix.

### b). GTF annotation file.
For the organism being analyzed, user need to provide an annotation file in gtf format to define the genomic coordinate of gene features. A good source to download those supporting files are [iGenome](https://support.illumina.com/sequencing/sequencing_software/igenome.html).

### c). The RADAR DataSet

The __List__ object used by the RADAR package to store the read counts 
and associated parameters, which will usually be represented in the code
here as an object `radar`.

We will use a toy data set to demonstrate how to start from bam file to count reads in consecutive bins and construct a RADAR dataset. There are 9 contols and 8 cases sample in the toy dataset. We name them by Ctl# and Case# in the follow demonstrations.  
Please note that user should set `bam_dir` to directory where bam files are saved instead of setting it to package assocated example file folder 
`system.file("extdata",package = "RADAR")`.  
```{r prepare, message=FALSE}
library(RADAR)
samplenames <- c( paste0("Ctl",1:9), paste0("Case",1:8) )
bam_dir <- "~/PoissonGamma_Method/data"
samplenames
```

Here we created a character vector `samplenames` to define the names of sample, which will later tells RADAR to look for **input** and **IP** file with prefix samplenames.  
Next we need to use `countReads()` function to 1) concatenate exons of each gene to get a "longest isoform" transcript of each gene. 2) Divide transcripts into consecutive bins of user defined width. 3) count reads mapped to each bin.  
**note1** The parameter `modification = "m6A"` tells the function to look for bam file of IP by the name `samplename.m6A.bam`. If the user named the IP sample as `samplename.IP.bam`, `modification` should be set to `modification = "IP`. We enable flexible seeting for IP sample in cases where user perform MeRIP seq on multiple modification with shared **INPUT** library.  
**note2** In Linux or MacOs, our package support multi-thread mode enabled by R package [doParallel](https://cran.r-project.org/web/packages/doParallel/index.html) . One can set the number of threads to use by `threads = n`.  
**note3** The default setting for bin width to slice the transcript is 50bp. One can set this by parameter `binSize = #bp `. For very shallow sequencing depth (e.g. less than 10M mappable reads per library), we recommand setting bin width to larger size such as 100bp to increase the number of reads countable in each bin.  
**note4** The parameter `shift` defines the number of nucleotide to shift from 5' end of reads to the center of RNA fragment. This number should be set to the half of your RNA fragment size in the sequencing library. In the example data, the RNA has been sonicated into ~150nt fragment before IP and library preparation.  
```{r count_read, warning=FALSE}
radar <- countReads(samplenames = samplenames,
                    gtf = "~/Database/genome/hg38/hg38_UCSC.gtf",
                    bamFolder = bam_dir,
                    modification = "m6A",
                    shift = 75,
                    outputDir = "test",
                    threads = 12,
                    binSize = 50
)
```


Now we have read count of consecutive 50bp bins for each sample. The count data is saved as a matrix object in the return List object `radar`. The count matrix can be returned by `radar$reads`. We can take a look at the content of the `radar` object.
```{r}
lapply(radar,head)
```

As shown, we can see `countReads()` function returns a List object of 6 elements. `RMD$reads` stores the read count matrix. `radar$binSize` stores the width of the bin to slice the transcript. `radar$geneModel` is a genomic range list object store the exons of each gene.  

### e). Normalization and Filtering  
After obtaining the `radar` object containing read count, we can then proceed to library normalization step. This step requires the user to define the grouping infomation by a vector parameter `X`.
```{r normalizeLibrary, warning=FALSE}
X <- c(rep("Ctl",9),rep("Case",8))
plot.new()
radar <- normalizeLibrary(radar,X)
```

In this step, the _bin-level_ read count of **input** is summarized into _gene-level_ read count. The size factor of **input** is calculated using "mean ratio method" implemented in _DESeq2_, which is then used to calculated normalized **input** _bin-level_ read count. To normalize the library size of **IP** samples, we use top-read-count bins of **IP** and corresponding **input** gene-level read count to compute the enrichment esimate of each sample. Under the assumption that samples in the same study have same IP efficiency, we normalize the **IP** by normalizing the estimated **IP efficiency**. As an sanity check, in addition to return the radar object with normalized read count, the `normalizeLibrary()` function also generate box plot of read count before and after normalizing **INPUT** and box plot of estimated enrichment before and after normalizing **IP**.  

In the returned radar object, there will be normalized gene-level **INPUT** count matrix stored in `radar$geneSum`, which is essentially RNA-seq read count matrix. The normalized _bin-level_ read count matrix will be `radar$norm.input`, `radar$norm.ip` respectively. The `radar` object will also contain size factor `radar$sizeFactor` used to normalize each library. 
```{r}
radar$sizeFactor
```
The grouping of the samples can be viewed by:
```{r}
radar$X
```
  
Next, we can use the _geneSum_ of **INPUT** to adjust for the variation of expression level of the **IP** read count. This step aims to account for the variation of **IP** read count attributed to variation of **gene expression level**.
```{r}
radar <- adjustExprLevel(radar)
```

This step add a new data `radar$ip_adjExpr` to the `radar` object. This matrix data are bin-level read count of **IP** adjusted for gene expression level variation. Then we will need to filter out bins of very low read count because under sampled locus can be strongly affected by technical variation. 
```{r}
radar <- filterBins(radar,minCountsCutOff = 15)
```

### e). Run Poisson Gamma Test  

Now we have the count matrix for testing differential methylation ready.  
To run the default PoissonGamma test, we can call the `diiIP()` function:
```{r, eval=FALSE}
radar <- diffIP(radar)
```
The code above run test on the effect of variable X defining the experimental group of samples, which is fine if the study is well designed and all known covariates are balanced. However, in most cases, samples might have been processed in more than one batches (especially when sample size is large) where batch effect and other covariate could contributed to the variation.  


### f). PCA analysis  

In order to check for unwanted variation, we can take the top 5000 bins ranked by count number (basically using the high read count bins) to plot PCA:
```{r}
top_bins <- radar$ip_adjExpr_filtered[order(rowMeans(radar$ip_adjExpr_filtered))[1:5000],]
plotPCAfromMatrix(top_bins,group = radar$X)
```
  
We can see that our Case and Control are not separated by the first two PCs. To check if the samples are separated by batches, we can label the samples by batch instead of by experimental group `X`.
```{r}
batch <- c("A","A","A","A","B","B","C","C","C","A","A","B","A","B","B","A","C")
plotPCAfromMatrix(top_bins,group = batch)
```
  
We can see that the samples are clearly separated by batches and we need to account for batch effect in the test. To incorporate **covariates** in study design, we need first define a covariate matrix table where categorical variable should be converted to (0,1) binary variable. 
```{r}
library(rafalib)
X2 <- as.fumeric(c("A","A","A","A","B","B","A","A","A","A","A","B","A","B","B","A","A"))-1 # batch as covariates
X3 <- as.fumeric(c("A","A","A","A","A","A","C","C","C","A","A","A","A","A","A","A","C"))-1 # batch as covariates
X4 <- as.fumeric(c("M","M","F","M","M","M","F","M","M","M","M","F","M","M","M","M","F"))-1 # sex as covariates
cov <- cbind(X2,X3,X4)

cov
```

Run the Poisson Gamma test with covariates to account for know source of variation. 
```{r, eval=FALSE}
radar <- diffIP(radar,Covariates = cov)
```  
or multithread mode 
```{r run_covariates, message=FALSE , echo=FALSE}
radar <-  diffIP_parallel(radar,Covariates = cov,exclude = NULL,thread = 10)
```

At the end of test, a histgram of p-value distribution will be plotted. The plot can be turned off by `diffIP(radar,Covariates = cov,plotPvalue = FALSE)`.  

If for some reasons you want to remove a few samples from the test, you can set index number of the samples you want to skip to exclude them in the Poisson Gamma test.
```
diffIP(radar,Covariates = cov,exclude = c())
```

In the `radar` object returned by the `diffIP()`, a mtrix of test estimators has been stored at 
```radar$all.est```.  

User can check the overall distribution of beta by
```{r}
hist(radar$all.est[,"beta1"],breaks = 50,main = "Distribution of effect size",xlab = "beta")
## In case of test without covariates, substitute "beta1" by "beta".
```

### g). Visualize the result by heatmap
To assess the pattern of variation of the MeRIP-seq data, we can plot the heatmap of normalized **IP** read counts adjusted for expression level.  
The belowing code shows how to plot heatmap on the radar object.
```{r}
plotHeatMap(radar)
```

In our example, as we determined earlier by PCA plot that batch effect is contibuting substantially to the variation of our data set. we need to plot heatmap on data regressing out covariates as determined earlier in the analysis. To do so, add covariates matrix to the `plotHeatMap()` function.
```{r}
plotHeatMap(radar,covariates = cov)
```

Since plotting too many bins willl look very messy, the function takes the top 5000 bins ranked by test p value. In another words, the heatmap generated here is a visualization the top differential methylated loci. From the heatmap we can see the methylation profile of `Ctl8` looks much closer to Cases than to Controls, which suggest this sample is an outlier.  
At this point, one can proceed to the next step to report the significantly differentially methylated bins to genomic location or repeat the analysis excluding the outlier sample. 

### h). Report the result

The count matrix we worked on above are on continuous bins on the transcript and the test was performed on each bin. To report the final result, we want to merge neighbouring bins that pass a significance cutoff because signals for neighbouring bins are assumed to come from the same locus. We merge the p-value of connecting bins by [fisher's method](https://en.wikipedia.org/wiki/Fisher%27s_method) and report the max beta from neighbouring bins. Besides, We also need to report the genomic location instead of on mRNA location of the bin that passed the cutoff. To do so, call the function:
```{r}
Diff_peak <- reportPoissonGammaMerge(radar,cutoff = 0.05)
```
The peak is reported at bed12 format with addition column for beta and p value.  
```{r}
head(Diff_peak)
```
### i). Visualize MeRIP-seq data by sequencing coverage plot.

One of the most straight forward visualization of MeRIP-seq data is the coverage plot. We implemented function to take a `radar` object and plot the mean/median coverage of IP and INPUT for two inferential groups. Here we demonstrate the function using a differential methylated gene reported in above sections.  
To use the plot function, we need to prepare an GTF annotation in GRanges format. This can be done by `GTF <- rtracklayer::import("path/to/gtf")`. This step only need to be done once in each R session. 
```{r plotCoverage, fig.width=15,fig.height=5}
GTF <- rtracklayer::import("~/Database/genome/hg38/hg38_UCSC.gtf")
plotGeneRADAR(radar,geneName = "SLC25A24", center = median, libraryType = "opposite",GTF = GTF)
```

**Note** The groupp label *Treat* and *Ctl* is automatically determined by the group vector `X` defined ealier (you can check by `radar$X`). The first factor in `X` is taken as *Ctl* and another group factor will be taken as *Treat*.  
**Note** The parameter `center = median` define the coverage to be group median. You can also set it to be `center = mean`.  

In some cases, the gene has very long introns that the exon regions becomes too narrow to see in the whole gene plot. We can call the same function with additional parameter `ZoomIn` to define a range in the gene to plot (Note `ZoomIn` range need to be within the gene). 

```{r plotCoverageZoom, fig.width=10,fig.height=5}
plotGeneRADAR(radar,geneName = "SLC25A24", center = median, libraryType = "opposite",GTF = GTF,ZoomIn = c(108200059,108200308 ) )
```

From the coverage plot we can see the in the "Treat" group, the m6A peak is significantly decreased. 


## 4. Model

### a). Poisson model  
 \begin{eqnarray*}
Y_{i} &\sim& Poi(\lambda_{i})  \\
\log(\lambda_{i} ) & = &   \mu_1 + X_i \beta +e_{i}
 \end{eqnarray*}

 We assume random effect $e_{i} \in \log\text{Gamma}(\psi, \psi)$, with scale parameter $\psi$ and mean equal to $1$. This is equivalent to $\lambda_{i} = e^{ \mu_1 + X_i \beta} w_i$ where $w_i \in \text{Gamma}(\psi, \psi)$ . \

\begin{eqnarray*}
 P( Y_i |  \Theta, - w_i) &=& \int e^{-w_i e^{  \mu_1 + X_i \beta}} \frac{(w_i e^{ \mu_1 + X_i \beta} )^{Y_i}}{Y_i!} \frac{\psi^\psi w_i^{\psi-1}e^{-\psi w_i}}{\Gamma(\psi)} d w_i\\ 
 &=& \frac{ \psi^\psi }{Y_i! } \frac{ \Gamma(Y_i+\psi) } {\Gamma(\psi) } \frac{ (e^{  \mu_1 + X_i \beta})^{Y_i}}{ (e^{  \mu_1 + X_i \beta} + \psi)^{Y_i + \psi}}
\end{eqnarray*}
  
 The marginal log likelihood of observing $\mathbf{Y}$  can be written as:
  \begin{eqnarray*}
\log L( \mathbf{Y} ) &=&  \sum_{i}^n   [ Y_i(  \mu_1 + X_i \beta) + \psi \log \psi + \log \Gamma(Y_i  + \psi) - \log Y_i! -  \log \Gamma( \psi) -  (Y_i  + \psi) \log( e^{  \mu_1 + X_i \beta} + \psi)] \\
 \end{eqnarray*}
 
### b). Estimation  
 We use gradient ascent algorithm to estimate the parameters, which involves the calculation of first derivatives.   
\begin{eqnarray*}
\frac{ \partial \log L( \mathbf{Y} ) }{ \partial \beta } &=& \sum_{i}^n  \ \Big[  Y_i  - \big (Y_i + \psi \big )\frac{e^{  \mu_1+ X_i \beta}}{ e^{  \mu_1 + X_i \beta} +\psi} \Big] X_i \\
\frac{ \partial \log L( \mathbf{Y} )  }{ \partial \psi} &=&  \sum_{i}^n  \Big [ \log \psi + 1 - \frac{Y_i +  \psi }{  e^{  \mu_1 + X_i \beta} +\psi} - \log ( e^{  \mu_1 + X_i \beta} +\psi ) \\
 &&+ \text{digamma} ( Y_i +  \psi)  -   \text{digamma} (  \psi)   \Big ] \\
\frac{ \partial \log L( \mathbf{Y} ) }{ \partial \mu _1} &=&   \sum_{i}^n  \Big[  Y_i - \big (Y_i + \psi \big ) \frac{e^{  \mu_1 + X_i \beta} }{e^{  \mu_1 + X_i \beta} + \psi}  \Big] 
 \end{eqnarray*}
 In each iteration, the parameters are updated through $ \Phi_{(t+1)} = \Phi_{(t)} + s_{(t+1)} \frac{ \partial \log( \mathbf{Y} ) }{ \partial \Phi}_ {|\Phi= \Phi_{(t)}}$.  The step size $s_{(t+1)}$ is determined by a line search algorithm. 
 
 
### c). Inference  
 We use log-likelihood ratio test to make the inference, where the null model is 
\begin{eqnarray*}
Y_{i} &\sim& Poi(\lambda_{i})  \\
\log(\lambda_{i} ) & = &   \mu_1 + e_{i}.
 \end{eqnarray*}


  
### d). Poisson model + covariates  
 \begin{eqnarray*}
Y_{i} &\sim& Poi(\lambda_{i})  \\
\log(\lambda_{i} ) & = &   \mu_1 + \mathbf{X_i} \mathbf{\beta} +e_{i}
 \end{eqnarray*}

 We assume random effect $e_{i} \in \log\text{Gamma}(\psi, \psi)$, with scale parameter $\psi$ and mean equal to $1$. This is equivalent to $\lambda_{i} = e^{ \mu_1 + \mathbf{X_i} \mathbf{\beta}} w_i$ where $w_i \in \text{Gamma}(\psi, \psi)$ . 

\begin{eqnarray*}
 P( Y_i |  \Theta, - w_i) &=& \int e^{-w_i e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}}} \frac{(w_i e^{ \mu_1 + \mathbf{X_i} \mathbf{\beta}} )^{Y_i}}{Y_i!} \frac{\psi^\psi w_i^{\psi-1}e^{-\psi w_i}}{\Gamma(\psi)} d w_i\\ 
 &=& \frac{ \psi^\psi }{Y_i! } \frac{ \Gamma(Y_i+\psi) } {\Gamma(\psi) } \frac{ (e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}})^{Y_i}}{ (e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}} + \psi)^{Y_i + \psi}}
\end{eqnarray*}
  
 The marginal log likelihood of observing $\mathbf{Y}$  can be written as:
  \begin{eqnarray*}
\log L( \mathbf{Y} ) &=&  \sum_{i}^n   [ Y_i(  \mu_1 + \mathbf{X_i} \mathbf{\beta} ) + \psi \log \psi + \log \Gamma(Y_i  + \psi) - \log Y_i! -  \log \Gamma( \psi) -  (Y_i  + \psi) \log( e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}} + \psi)] \\
 \end{eqnarray*}
 
#### Estimation
 We use gradient ascent algorithm to estimate the parameters, which involves the calculation of first derivatives.   
\begin{eqnarray*}
\frac{ \partial \log L( \mathbf{Y} ) }{ \partial \beta_k } &=& \sum_{i}^n  \ \Big[  Y_i  - \big (Y_i + \psi \big )\frac{e^{  \mu_1+ \mathbf{X_i} \mathbf{\beta}}}{ e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}} +\psi} \Big] X_{ik} \\
\frac{ \partial \log L( \mathbf{Y} )  }{ \partial \psi} &=&  \sum_{i}^n  \Big [ \log \psi + 1 - \frac{Y_i +  \psi }{  e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta} } +\psi} - \log ( e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}} +\psi ) \\
 &&+ \text{digamma} ( Y_i +  \psi)  -   \text{digamma} (  \psi)   \Big ] \\
\frac{ \partial \log L( \mathbf{Y} ) }{ \partial \mu _1} &=&   \sum_{i}^n  \Big[  Y_i - \big (Y_i + \psi \big ) \frac{e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}} }{e^{  \mu_1 + \mathbf{X_i} \mathbf{\beta}} + \psi}  \Big] 
 \end{eqnarray*}
 In each iteration, the parameters are updated through $ \Phi_{(t+1)} = \Phi_{(t)} + s_{(t+1)} \frac{ \partial \log( \mathbf{Y} ) }{ \partial \Phi}_ {|\Phi= \Phi_{(t)}}$.  The step size $s_{(t+1)}$ is determined by a line search algorithm. 
 
 
#### Inference 
We use log-likelihood ratio test to make the inference for each $\beta_k$, where the null model is 
\begin{eqnarray*}
Y_{i} &\sim& Poi(\lambda_{i})  \\
\log(\lambda_{i} ) & = &   \mu_1 + \sum_{j \neq k}^{K} X_{ij} \beta_j + e_{i}.
 \end{eqnarray*}
We use observed fisher information to make the inference for all parameters. The calculation involves all second derivatives and partial derivatives.   
 \begin{eqnarray*}
\frac{ \partial^2\log L( \mathbf{Y} )}{\partial \beta_k^2} &=&  -  \sum_{i}^n   \frac{e^{  \mu_1 +  \mathbf{X_i} \mathbf{\beta} }}{ \big ( e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi \big ) ^2}    \psi   \big (Y_i + \psi \big ) X_{ik}^2  \\
\frac{ \partial^2\log L( \mathbf{Y} )}{\partial \beta_k \beta_k'} &=&  -  \sum_{i}^n   \frac{e^{  \mu_1 +  \mathbf{X_i} \mathbf{\beta} }}{ \big ( e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi \big ) ^2}    \psi   \big (Y_i + \psi \big )  X_{ik} X_{ik'}  \\
\frac{ \partial^2 \log L( \mathbf{Y} ) }{\partial \beta_k  \partial \mu_1} &=&  - \sum_{i}^n   \frac{e^{  \mu_1 +  \mathbf{X_i} \mathbf{\beta} }}{ \big ( e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi \big ) ^2}    \psi   \big (Y_i + \psi \big ) X_{ik}  \\
\frac{ \partial^2 \log L( \mathbf{Y} ) }{\partial \beta_k  \partial \psi} &=&   \sum_{i}^n  \Big[   \frac{e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta} } }{ \big ( e^{  \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi \big ) ^2}  \big (Y_i + \psi \big ) X_{ik}  - \frac{e^{   \mu_1 +  \mathbf{X_i} \mathbf{\beta} } }{ e^{   \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi } X_{ik} \Big]\\
\frac{ \partial^2 \log L( \mathbf{Y} ) }{\partial \mu_1^2} &=& - \sum_{i}^n   \frac{e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta}} }{ \big ( e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi \big ) ^2}  \psi  \big (Y_i + \psi \big )  \\
\frac{ \partial^2 \log L( \mathbf{Y} ) }{\partial \mu_1 \partial \psi} &=& \sum_{i}^n  \Big[   \frac{e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta} } }{ \big ( e^{  \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi \big ) ^2}  \big (Y_i + \psi \big ) -\frac{e^{   \mu_1 +  \mathbf{X_i} \mathbf{\beta} } }{ e^{   \mu_1 +  \mathbf{X_i} \mathbf{\beta} }  + \psi } \Big] \\
\frac{ \partial^2 \log L( \mathbf{Y} ) }{\partial \psi^2} &=&  \sum_{i}^n \Big[  \frac{1}{\psi}  - \frac{2}{e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta} } + \psi}  +  \frac{Y_i + \psi }{ \big ( e^{ \mu_1 +  \mathbf{X_i} \mathbf{\beta}} + \psi \big ) ^2} + \text{trigamma}(Y_i + \psi) -  \text{trigamma}(\psi)  \Big] \\
\end{eqnarray*}
 




## Acknowledgments

We have benefited in the development of RADAR from the help and
feedback of many individuals, including but not limited to: 




## Session info

```{r sessionInfo}
sessionInfo()
```

## References

[1] Love, M. I.; Huber, W.; & Anders, S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology. 2014 15(12), 550. 

